src/servers/src/pipeline.rs

```diff
@@ -0,0 +1,224 @@
+    db: &str,
+    transform_timer: &Instant,
+    is_top_level: bool,
+) -> Result<()> {
+    let r = pipeline
+        .exec_mut(intermediate_state)
+        .inspect_err(|_| {
+            if is_top_level {
+                METRIC_HTTP_LOGS_TRANSFORM_ELAPSED
+                    .with_label_values(&[db, METRIC_FAILURE_VALUE])
+                    .observe(transform_timer.elapsed().as_secs_f64());
+            }
+        })
+        .context(PipelineTransformSnafu)
+        .context(PipelineSnafu)?;
+
+    match r {
+        PipelineExecOutput::Transformed(row) => {
+            transformed.push(row);
+        }
+        PipelineExecOutput::DispatchedTo(dispatched_to) => {
+            if let Some(values) = dispatched.get_mut(&dispatched_to) {
+                values.push(intermediate_state.clone());
+            } else {
+                dispatched.insert(dispatched_to, vec![intermediate_state.clone()]);
+            }
+        }
+    }
+
+    Ok(())
+}
+
+/// Never call this on `GreptimeIdentityPipeline` because it's a real pipeline
+pub async fn get_pipeline(
+    pipeline_def: PipelineDefinition,
+    handler: &PipelineHandlerRef,
+    query_ctx: &QueryContextRef,
+) -> Result<Arc<Pipeline<GreptimeTransformer>>> {
+    match pipeline_def {
+        PipelineDefinition::Resolved(pipeline) => Ok(pipeline),
+        PipelineDefinition::ByNameAndValue((name, version)) => {
+            handler
+                .get_pipeline(&name, version, query_ctx.clone())
+                .await
+        }
+        _ => {
+            unreachable!("Never call get_pipeline on identity.")
+        }
+    }
+}
+
+pub(crate) async fn run_pipeline(
+    state: &PipelineHandlerRef,
+    pipeline_definition: PipelineDefinition,
+    values: PipelineExecInput,
+    table_name: String,
+    query_ctx: &QueryContextRef,
+    db: &str,
+    is_top_level: bool,
+) -> Result<Vec<RowInsertRequest>> {
+    if matches!(
+        pipeline_definition,
+        PipelineDefinition::GreptimeIdentityPipeline
+    ) {
+        let table = state
+            .get_table(&table_name, query_ctx)
+            .await
+            .context(CatalogSnafu)?;
+        pipeline::identity_pipeline(values, table)
+            .map(|rows| {
+                vec![RowInsertRequest {
+                    rows: Some(rows),
+                    table_name,
+                }]
+            })
+            .context(PipelineTransformSnafu)
+            .context(PipelineSnafu)
+    } else {
+        let pipeline = get_pipeline(pipeline_definition, state, query_ctx).await?;
+
+        let transform_timer = std::time::Instant::now();
+        let mut intermediate_state = pipeline.init_intermediate_state();
+
+        let mut transformed = Vec::with_capacity(values.len());
+        let mut dispatched: BTreeMap<DispatchedTo, Vec<Vec<pipeline::Value>>> = BTreeMap::new();
+
+        match values {
+            PipelineExecInput::Original(array) => {
```

Comment:
`let mut intermediate_state = pipeline.init_intermediate_state();`
Can we move this line under 
`PipelineExecInput::Original(array) => {`

Reply:
good catch!